### PR TITLE
docs: docs changes for searcher context removal

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

a few docs updates from searcher context removal.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] New features have been approved by the corresponding PM
- [ ] User-facing API changes have the "User-facing API Change" label
- [ ] Release notes have been added as a separate file under `docs/release-notes/`
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses have been included for new code which was copied and/or modified from any external code

[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/model-dev-guide/api-guides/apis-howto/deepspeed/deepspeed.rst [hunk 1]
```diff
@@ -365,6 +365,241 @@ profiling batches 3 and 4.
...
+           )
+
+
+   if __name__ == "__main__":
+       # Configure logging
+       logging.basicConfig(level=logging.INFO, format=det.LOG_FORMAT)
+       main()
+
+You can run this Python script directly (``python3 train.py``), or in a Jupyter notebook. This code
+will train for ten epochs, and checkpoint and validate every 100 batches.
```
Comment (rb-determined-ai) [inline docs/model-dev-guide/api-guides/apis-howto/deepspeed/deepspeed.rst:462]: I think the comma is extraneous, but also the phrasing is odd.

Maybe: "This code will train for ten epochs, checkpointing and validating every 100 batches"?

File: docs/model-dev-guide/api-guides/apis-howto/deepspeed/deepspeed.rst [hunk 2]
```diff
@@ -365,6 +365,241 @@ profiling batches 3 and 4.
...
+
+.. code:: diff
+
+   import deepspeed
+
+   def main():
+   +     # Initialize distributed backend before det_ds.init()
+   +     deepspeed.init_distributed()
+   +     # Set flag used by internal PyTorch training loop
+   +     os.environ["DET_MANUAL_INIT_DISTRIBUTED"] = "true"
```
Comment (rb-determined-ai) [inline docs/model-dev-guide/api-guides/apis-howto/deepspeed/deepspeed.rst:478]: I'm sorry, what?

If we can't fix this before the release ships, then I vote we remove this section and add it back later when we fix it.

This is basically a bug, and we're documenting it here rather than fixing it.

File: docs/model-dev-guide/api-guides/apis-howto/deepspeed/deepspeed.rst [hunk 3]
```diff
@@ -365,6 +365,241 @@ profiling batches 3 and 4.
...
+Once you are satisfied with the results of training the model locally, you submit the code to a
+cluster. This example allows for distributed training locally and on cluster without having to make
+code changes.
+
+Example workflow of frequent iterations between local debugging and cluster deployment:
+
+.. code:: diff
+
+    def main():
+   +   local = det.get_cluster_info() is None
```
Comment (rb-determined-ai) [inline docs/model-dev-guide/api-guides/apis-howto/deepspeed/deepspeed.rst:523]: do:

```python
info = det.get_cluster_info()
if info is not None:
    # Local: configure local distributed training.
```

Something like `local = det.get_cluster_info() is None` only makes sense if you want the code to be self-documenting to people who don't really know the codebase.  It doesn't work from a mypy perspective, I don't think, because mypy can't assume that `get_cluster_info()` is going to return the same thing every time.

But you have a comment about being local anyway, so write the code normally and let the comment be the explanation.

File: docs/model-dev-guide/api-guides/apis-howto/deepspeed/deepspeed.rst [hunk 4]
```diff
@@ -365,6 +365,241 @@ profiling batches 3 and 4.
...
+             trial = DCGANTrial(train_context)
+             trainer = det_ds.Trainer(trial, train_context)
+             trainer.fit(
+                 max_length=pytorch.Epoch(11),
+                 checkpoint_period=pytorch.Batch(100),
+                 validation_period=pytorch.Batch(100),
+   +             latest_checkpoint=latest_checkpoint,
+             )
+
+To run Trainer API solely on-cluster, the code is much simpler:
```
Comment (rb-determined-ai) [inline docs/model-dev-guide/api-guides/apis-howto/deepspeed/deepspeed.rst:549]: drop the word "much", not only is it fluff, it's fluff that makes our API look worse.

File: docs/release-notes/searcher-context-removal.rst [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+   and are now being removed. Users are encouraged to use a preset searcher, which can be easily
+   :ref:`configured <experiment-configuration_searcher>` for any experiment.
+
+**New Features**
+
+-  API: introduce ``keras.DeterminedCallback``, a new high-level training API for TF Keras that
+   integrates Keras training code with Determined through a single :ref:`Keras Callback
+   <api-keras-ug>`.
+
+-  API: introduce ``deepspeed.Trainer``, a new high-level training API for DeepSpeedTrial that
```
Comment (MikhailKardash) [inline docs/release-notes/searcher-context-removal.rst:21]: One of the changes to `DeepSpeedTrialContext` was to remove `det.EnvContext` from it. This used to be accessible via `context.env`. I mentioned this as a potential breaking change for some users, who may have to move to the `context.get_experiment_config()` and `context.get_hparams()` methods.

File: docs/reference/experiment-config-reference.rst [hunk 6]
```diff
@@ -319,25 +319,6 @@ While debugging, the logger will display lines highlighted in blue for easy iden
  Validation Policy
 *******************
 
-.. _experiment-config-min-validation-period:
```
Comment (rb-determined-ai) [inline docs/reference/experiment-config-reference.rst:322]: uh, I should probably tell you that @tara-det-ai told me we should mark it as deprecated but not delete it from the docs.

This is contrary to my preference, which is to let users look in old docs if they want to use old features.

But also it makes sense in this case because we don't have any deprecation warnings anywhere in the system for these fields; users would only know by looking at the docs.

Also this means we need to revert some of my docs deletions from the searcher-context-removal branch, which I realize now that I forgot to do :disappointed:.  It would be the removals from this file, I think.

That said, the other places where you remove things like "use min_validation_period to control validations" should be removed.  But the actual reference here should not.

File: docs/model-dev-guide/api-guides/apis-howto/deepspeed/deepspeed.rst (hunk 2 above)
Comment (azhou-determined) [reply docs/model-dev-guide/api-guides/apis-howto/deepspeed/deepspeed.rst]: well, this is copied from pytorch-ug, where we document similar behavior. which isn't a bug, but a hackily-supported use case.

i'm pretty sure this isn't needed here though, since either we're in local mode, we init the dist backend and pass it into the ds.init(), or we're on cluster and it's None.

so i removed it, cc: @MikhailKardash

File: docs/release-notes/searcher-context-removal.rst (hunk 5 above)
Comment (MikhailKardash) [reply docs/release-notes/searcher-context-removal.rst:21]: Actually, should probably just tell users it's gone completely and they should read config from the cluster info.

File: docs/release-notes/searcher-context-removal.rst [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+   allows for Python-side training loop configurations and includes support for local training.
+
+**Deprecations**
+
+-  Experiment Config: the ``max_length`` field of the searcher configuration section has been
+   deprecated for all experiments and searchers. Users are expected to configure the desired
+   training length directly in training code.
+
+-  Experiment Config: the ``optimizations`` config has been deprecated. Please see :ref:`Training
```
Comment (rb-determined-ai) [inline docs/release-notes/searcher-context-removal.rst:30]: Are all these deprecations mirrored in the rest of the docs?  I didn't see optimizations touched in the experiment config reference, for instance.

File: examples/deepspeed/dcgan/model.py [hunk 8]
```diff
@@ -46,7 +46,7 @@ def __init__(self, context: det_ds.DeepSpeedTrialContext,
         self.discriminator = self.context.wrap_model_engine(discriminator)
         self.fixed_noise = self.context.to_device(
             torch.randn(
-                self.context.train_micro_batch_size_per_gpu, self.hparams["noise_length"], 1, 1
```
Comment (rb-determined-ai) [inline examples/deepspeed/dcgan/model.py:49]: was this a bug, or is this a breaking change to the API?

File: harness/determined/pytorch/_trainer_utils.py [hunk 9]
```diff
@@ -87,15 +83,25 @@ def _divides(self, steps: int) -> bool:
 
 class Epoch(TrainUnit):
     """
-    Epoch step type (e.g. Epoch(1) defines 1 epoch)
+    Defines an Epoch unit for specifying length to PyTorch trainers.
+
+    Epoch(int) values are treated as periods, e.g. Epoch(100) will checkpoint/validate every 100
+    epochs.
+    Epoch(collections.abc.Container) values are treated as schedules, e.g. Epoch(1,5,10) will
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_trainer_utils.py:90]: Uh, shouldn't this be `Epoch([1,5,10])` instead?  Same with the Batch docstring

File: docs/model-dev-guide/api-guides/apis-howto/deepspeed/deepspeed.rst (hunk 2 above)
Comment (rb-determined-ai) [reply docs/model-dev-guide/api-guides/apis-howto/deepspeed/deepspeed.rst]: If we are documenting internal flags, that is a bug.  It's also an oxymoron, since documenting it means it's not really internal, just a weirdly bad public API.

File: docs/release-notes/searcher-context-removal.rst (hunk 5 above)
Comment (rb-determined-ai) [reply docs/release-notes/searcher-context-removal.rst:21]: This shouldn't technically be breaking because it was never documented, so it was not technically public.

But it's so old that it wasn't prefixed with `_` because we didn't used to be good about doing that.

Comment (rb-determined-ai) [reply docs/release-notes/searcher-context-removal.rst:21]: If it wasn't documented before, we don't need to document that it is gone, imo.

File: docs/model-dev-guide/api-guides/apis-howto/deepspeed/deepspeed.rst (hunk 2 above)
Comment (azhou-determined) [reply docs/model-dev-guide/api-guides/apis-howto/deepspeed/deepspeed.rst]: "it's a feature, not a bug!"

lol, IIRC at the time of landing pytorch trainer, we decided to do this because we wanted to document a shiny new local training capability but didn't have time to figure out a good way to do it. 

you're right that it's a bad public API. i've put it on my todo list, and i'll fix it next week. it's not related to this PR or this feature, though.

File: examples/deepspeed/dcgan/model.py (hunk 8 above)
Comment (azhou-determined) [reply examples/deepspeed/dcgan/model.py:49]: both? it was a bug because it was a breaking change. added to release notes.

File: docs/reference/experiment-config-reference.rst (hunk 6 above)
Comment (azhou-determined) [reply docs/reference/experiment-config-reference.rst:322]: welp. ok then. i added them all back.

File: docs/release-notes/searcher-context-removal.rst (hunk 5 above)
Comment (MikhailKardash) [reply docs/release-notes/searcher-context-removal.rst:25]: The reason I reference it is because our examples had code that used it (`gpt_neox` before the Trainer API rewrites, for example)

File: docs/model-dev-guide/api-guides/apis-howto/deepspeed/deepspeed.rst [hunk 10]
```diff
@@ -365,6 +365,237 @@ profiling batches 3 and 4.
...
+   +   if info is None:
+   +       # Local: configure local distributed training.
+   +       deepspeed.init_distributed()
+   +       distributed_context = core.DistributedContext.from_deepspeed()
+   +       latest_checkpoint = None
+   +   else:
+   +       # On-cluster: Determined will automatically detect distributed context.
+   +       distributed_context = None
+   +       # On-cluster: configure the latest checkpoint for pause/resume training functionality.
+   +       latest_checkpoint = det.get_cluster_info().latest_checkpoint
```
Comment (rb-determined-ai) [inline docs/model-dev-guide/api-guides/apis-howto/deepspeed/deepspeed.rst:531]: reuse `info`

File: docs/reference/experiment-config-reference.rst (hunk 6 above)
Comment (rb-determined-ai) [reply docs/reference/experiment-config-reference.rst:322]: thanks, sorry.

File: examples/deepspeed/dcgan/model.py (hunk 8 above)
Comment (rb-determined-ai) [reply examples/deepspeed/dcgan/model.py:49]: This is the only reason we are breaking existing DeepSpeedTrials?  Like, we're deprecating old paths, yes, but we're only actually breaking user code for this?

Can we include a `@property`-style getter that makes this _not_ a breaking change?

something like:

```python
    @property
    def train_micro_batch_size_per_gpu(self) -> int:
        warnings.warn(
            "DeepSpeedTrialContext.train_micro_batch_size_per_gpu has been deprecated in "
            "Determined 0.38.0; please use the context.get_train_micro_batch_size_per_gpu() getter "
            "instead."
            FutureWarning,
            stacklevel=2,
        )
        return self.get_train_micro_batch_size_per_gpu()
```

File: docs/release-notes/searcher-context-removal.rst [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+   training loops now automatically report ``batches`` and ``epochs`` with metrics, which you can
+   use as your ``time_metric``. ASHA experiments without this modification will no longer run.
+
+-  Custom Searchers: all custom searchers including DeepSpeed Autotune were deprecated in ``0.36.0``
+   and are now being removed. Users are encouraged to use a preset searcher, which can be easily
+   :ref:`configured <experiment-configuration_searcher>` for any experiment.
+
+-  DeepSpeed: the ``num_micro_batches_per_slot`` and ``train_micro_batch_size_per_gpu`` attributes
+   on ``DeepSpeedContext`` have been replaced with ``get_train_micro_batch_size_per_gpu()`` and
+   ``get_num_micro_batches_per_slot()``.
```
Comment (rb-determined-ai) [inline docs/release-notes/searcher-context-removal.rst:17]: no longer breaking

**Review by rb-determined-ai — APPROVED** [review]
awesome!
